prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -0,0 +1,197 @@
+    if pd.isna(generator["lat"]) or pd.isna(generator["lon"]):
+        # No location available
+        try:
+            matching_subs = substation_groupby.get_group(lookup_params)
+            return matching_subs.index[0]
+        except KeyError:
+            return pd.NA
+    try:
+        # This ZIP code contains substations, this block will execute successfully
+        matching_subs = substation_groupby.get_group(lookup_params)
+    except KeyError:
+        # If this ZIP code does not contain substations, this block will execute, and
+        # we select a set of 'nearby' substations
+        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
+        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
+        try:
+            matching_subs = pd.concat(
+                [
+                    substation_groupby.get_group((generator.loc["interconnect"], z))
+                    for z in zip_range_strings
+                    if (generator.loc["interconnect"], z) in substation_groupby.groups
+                ]
+            )
+        except ValueError:
+            # If no matching subs within the given interconnection and ZIPs, give up
+            return pd.NA
+    distance_to_subs = matching_subs.apply(
+        lambda x: haversine((x.LATITUDE, x.LONGITUDE), (generator.lat, generator.lon)),
+        axis=1,
+    )
+    return distance_to_subs.idxmin()
+
+
+def map_generator_to_bus_by_sub(generator, bus_groupby):
+    """Determine a likely bus for a generator to be connected to, based on the bus with
+    the lowest voltage within the generator's specified substation.
+
+    :param pandas.Series generator: one generating unit from data frame.
+    :param pandas.GroupBy bus_groupby: data frame of buses, grouped by substation ID.
+    """
+    if pd.isna(generator.sub_id):
+        return pd.NA
+    else:
+        return bus_groupby.get_group(generator.sub_id)["baseKV"].idxmin()
+
+
+def add_locations_to_generators(generators, power_plants):
+    """Add 'lat' and 'lon' columns to ``generators``, based on matching plant codes in
+    ``power_plants``. Non-matched generating units will get NaN values.
+
+    :param pandas.DataFrame generators: data frame of generating units
+        (will be modified with new columns).
+    :param pandas.DataFrame power_plants: data frame of power plants
+        (locations come from here).
+    """
+    plant_locations = pd.DataFrame(
+        power_plants[["LATITUDE", "LONGITUDE"]],
+        index=power_plants.index.astype("str"),
+    )
+    filtered_generators = generators.query("not PLANT_CODE.isnull()")
+    parsed_plant_codes = filtered_generators.PLANT_CODE.astype("int").astype("str")
+    generators["lat"] = parsed_plant_codes.map(plant_locations["LATITUDE"])
+    generators["lon"] = parsed_plant_codes.map(plant_locations["LONGITUDE"])
+
+
+def map_generators_to_interconnects(generators, power_plants):
```

Comment:
I think the description for this could be slightly modified - it sounds like the intent is to modify `generators` but from the way it's used, it is creating the interconnect column which is assigned within `build_plant`.

Reply:
You're right, that was how the function was originally written before I refactored it. Will fix.

Reply:
This function was made obsolete by starting from EIA data, which have NERC regions, which map cleanly to interconnects.